sanitizers/detectors/template_injection.go

```diff
@@ -2,37 +2,53 @@ package detectors
 
 import (
 	"errors"
+	htmlTemplate "html/template"
 	"strings"
+	textTemplate "text/template"
 	"text/template/parse"
 
 	"github.com/CodeIntelligenceTesting/gofuzz/sanitizers/fuzzer"
+	"github.com/CodeIntelligenceTesting/gofuzz/sanitizers/reporter"
 )
 
 const evilTemplateAction = "{{ .EvilAction }}"
 
-var TemplateInjectionError = errors.New("template injection error")
+var TemplateInjectionError = errors.New("Template injection error")
 
-type TemplateInjection struct {
-	id   int         // numeric identifier to distinguish between the detectors for the various call sites
-	tree *parse.Tree // representation of the parsed template
+var _ Detector = (*templateInjection)(nil)
+
+type templateInjection struct {
+	id    int
+	tree  *parse.Tree
+	vargs []any
 }
 
-// Make sure that the template injection detector implements the Detector interface
-var _ Detector = (*TemplateInjection)(nil)
+var TemplateInjection = templateInjection{}
 
-func (ti *TemplateInjection) Detect() error {
+func (ti *templateInjection) Detect() {
 	tmplText := ti.tree.Root.String()
 	if strings.Contains(tmplText, evilTemplateAction) {
-		return TemplateInjectionError
-	} else {
-		fuzzer.GuideTowardsContainment(tmplText, evilTemplateAction, ti.id)
+		reporter.ReportFindingf("%s", TemplateInjectionError.Error())
+		return
 	}
-	return nil
+	fuzzer.GuideTowardsContainment(tmplText, evilTemplateAction, ti.id)
+}
+
+func (ti *templateInjection) New(id int, ttype interface{}, vargs ...any) *templateInjection {
+	return &templateInjection{id, GetTree(ttype), vargs}
 }
 
-func NewTemplateInjection(id int, tree *parse.Tree) *TemplateInjection {
-	return &TemplateInjection{
-		id:   id,
-		tree: tree,
+func GetTree(ttype interface{}, args ...any) *parse.Tree {
```

Comment:
Could you explain why do we need an interface parameter here? What where the limitations of the previous approach of providing a `*parse.Tree` that represents the template being executed?

Reply:
Same reasoning as before with `GetCommand`, as in that this allowed us to just dump in a `*textTemplate.Template` or a `*htmlTemplate.Template` with extra args (if needed), that would have called `Lookup` in cases were the extra args were provided, again trying to abstract away as much as possible from the hook implementation. Reverted this.